proposals/2021/07-22_log-streaming.md

```diff
@@ -0,0 +1,306 @@
+| **Reviewers** | Neal.Coleman, David.May, Emmanuel.Meinen, Kelly.Merrick, David.Vader, Matthew.Fevold |
+| **Date**      | July 22nd, 2021                                                                      |
+| **Status**    | Reviewed                                                                             |
+
+<!--
+If you're already working with someone, please add them to the proper author/reviewer category.
+
+If not, please leave the reviewer category empty and someone from the Vela team will assign it to themself.
+
+Here is a brief explanation of the different proposal statuses:
+
+1. Reviewed: The proposal is currently under review or has been reviewed.
+
+2. Accepted: The proposal has been accepted and is ready for implementation.
+
+3. In Progress: An accepted proposal is being implemented by actual work.
+
+NOTE: The design is subject to change during this phase.
+
+4. Cancelled: While or before implementation the proposal was cancelled.
+
+NOTE: This can happen for a multitude of reasons.
+
+5. Complete: This feature/change is implemented.
+-->
+
+## Background
+
+<!--
+This section is intended to describe the new feature, redesign or refactor.
+-->
+
+**Please provide a summary of the new feature, redesign or refactor:**
+
+<!--
+Provide your description here.
+-->
+
+This enhancement will enable viewing logs for a service/step in near real-time.
+
+Currently, when watching logs for a service/step they aren't provided in real-time.
+
+Instead, we use a buffer mechanism for controlling how logs are published:
+
+1. service/step starts running on a worker producing logs
+2. logs that are produced from the service/step are pushed to a buffer
+3. if the buffer exceeds `1000` bytes
+   * publish the logs from the buffer via API call to the server
+   * flush the buffer so we can push more logs to it from the service/step
+4. circle back to number 2 until the service/step is complete
+5. once the service/step is complete, publish remaining logs from the buffer
+
+The end-behavior produced by this method is the logs appear in delayed chunks.
+
+**Please briefly answer the following questions:**
+
+1. Why is this required?
+
+<!-- Answer here -->
+
+* provide compatible functionality with existing CI solutions
+* improve user experience when viewing logs for a service/step
+* improve ability to troubleshoot pipelines by seeing what parts take longer
+
+2. If this is a redesign or refactor, what issues exist in the current implementation?
+
+<!-- Answer here -->
+
+The current implementation leaves a lot to be desired for user experience.
+
+With the logs appearing in delayed chunks, it's difficult to troubleshoot pipelines.
+
+This behavior can make processes appear to be "stuck" or "hung" when inspecting the logs.
+
+This can make it almost impossible to determine if something is running or how long it takes to run.
+
+3. Are there any other workarounds, and if so, what are the drawbacks?
```

Comment:
I think we could turn down the bytes from 1000 -> some smaller number... might also give a faux feeling of real time streaming? Not that I'm recommending that, just _another_ option :)

Reply:
I'll update the proposal to contain that information 👍

Reply:
This is done 👍